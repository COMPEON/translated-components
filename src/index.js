import React from 'react'
import kebabCase from 'lodash.kebabcase'
import transform from 'lodash.transform'
import isString from 'lodash.isstring'
import isNumber from 'lodash.isnumber'
import IntlFormat from 'intl-messageformat'

const DEFAULT_LOCALE = 'de_DE'

const CURRENCY_BY_REGION = {
  AT: 'EUR',
  CH: 'CHF',
  DE: 'EUR',
  FR: 'EUR',
  GB: 'GPB',
  US: 'USD'
}

const passThrough = value => value

const {
  Provider: TranslationProvider,
  Consumer: TranslationConsumer
} = React.createContext(DEFAULT_LOCALE)

const withTranslation = ({ translations, params = {}, mapTranslationsToProps = passThrough, format = {} }) => {
  const initializedTranslations = initializeTranslations(translations, format)

  const wrapComponent = Component => props => {
    const translationsForLocale = locale => translateWithDefaults({
      translations: initializedTranslations,
      locale: locale || DEFAULT_LOCALE,
      reducer: templateReducer(applyParamFunctions(props, params))
    })

    return (
      <TranslationConsumer>
        { locale =>
          <Component
            {...props}
            {...mapTranslationsToProps(translationsForLocale(locale), props)}
          />
        }
      </TranslationConsumer>
    )
  }

  return wrapComponent
}

const moneyFormat = locale => ({
  number: {
    money: {
      currency: CURRENCY_BY_REGION[locale.slice(-2)],
      minimumFractionDigits: 0,
      style: 'currency'
    }
  }
})

const applyIntlToTranslations = (translations, locale, format) => (
  transform(translations, (result, entry, key) => {
    result[key] = new IntlFormat(entry, kebabCase(locale), { ...moneyFormat(locale), ...format })
  })
)

const initializeTranslations = (translations, format) => (
  transform(translations, (result, localeTranslations, locale) => {
    result[locale] = applyIntlToTranslations(localeTranslations, locale, format)
  })
)

const templateReducer = (values) => (result, v, k) => {
  result[k] = v.format(values)
  return result
}

const cleanParams = params => (
  transform(params, (result, v, k) => {
    if (isString(v) || isNumber(v)) result[k] = v
  })
)

const applyParamFunctions = (props, params) => (
  transform(params, (result, fn, k) => {
    result[k] = fn(props)
  }, cleanParams(props))
)

const translateWithDefaults = ({ locale, reducer, translations = {} }) => (
  transform({
    ...translations[DEFAULT_LOCALE],
    ...translations[locale]
  }, reducer)
)

export { TranslationProvider, withTranslation }
