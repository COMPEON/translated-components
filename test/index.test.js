import React from 'react'
import { mount } from 'enzyme'
import { TranslationProvider, withTranslation } from '../src'

const SomeTestComponent = ({
  notToBeTranslated = 100,
  title = 'a default title',
  'kebab-label': kebabLabel = 'a default kebab label',
  message = 'a default message'
}) => (
  <React.Fragment>
    <img width={notToBeTranslated} />
    <h1>{title}</h1>
    <h2>{kebabLabel}</h2>
    <p>{message}</p>
  </React.Fragment>
)

const translations = {
  de_DE: {
    title: 'Dies ist ein fester Titel',
    'kebab-label': 'Du, du hast.',
    message: 'Eine Schweinshaxe, bitte.'
  },
  de_AT: {
    title: 'Dies ist ein {customTitle}',
    'kebab-label': 'Du hast {euros, number, money} Euros!',
    message: 'Hier wird nie etwas anderes stehen.'
  },
  en_GB: {
    title: 'Good day, Sir.'
  },
  en_US: {
    message: 'Let\'s eat {numBurgers, plural, =0 {no burgers} =1 {one burger} other {# burgers}} today.'
  },
  fr_FR: {
    title: 'This is going to be {interpolated}'
  }
}

const params = {
  interpolated: ({ interpolated = "" }) => `really ${interpolated.toUpperCase()}!`
}


describe('Translate components', () => {
  const TranslatedTestComponent = withTranslation({translations, params})(SomeTestComponent)

  const subject = ({locale, ...rest}) => (
    mount(
      <TranslationProvider value={locale}>
        <TranslatedTestComponent {...rest} />
      </TranslationProvider>
    )
  )
  it('does not touch non-translation props', () => {
    expect(subject({notToBeTranslated: 99})).toMatchSnapshot()
  })

  it('passes default `de_DE` translated strings to wrapped components', () => {
    expect(subject({})).toMatchSnapshot()
  })

  it('passes translated strings alongside default fallbacks to wrapped components', () => {
    expect(subject({locale: 'en_GB'})).toMatchSnapshot()
  })
})