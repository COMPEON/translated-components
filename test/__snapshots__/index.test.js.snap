// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Translate components does not touch non-translation props 1`] = `
<Component
  notToBeTranslated={99}
>
  <SomeTestComponent
    kebab-label="Du, du hast."
    message="Eine Schweinshaxe, bitte."
    notToBeTranslated={99}
    title="Dies ist ein fester Titel"
  >
    <img
      width={99}
    />
    <h1>
      Dies ist ein fester Titel
    </h1>
    <h2>
      Du, du hast.
    </h2>
    <p>
      Eine Schweinshaxe, bitte.
    </p>
  </SomeTestComponent>
</Component>
`;

exports[`Translate components passes default \`de_DE\` translated strings to wrapped components 1`] = `
<Component>
  <SomeTestComponent
    kebab-label="Du, du hast."
    message="Eine Schweinshaxe, bitte."
    title="Dies ist ein fester Titel"
  >
    <img
      width={100}
    />
    <h1>
      Dies ist ein fester Titel
    </h1>
    <h2>
      Du, du hast.
    </h2>
    <p>
      Eine Schweinshaxe, bitte.
    </p>
  </SomeTestComponent>
</Component>
`;

exports[`Translate components passes translated strings alongside default fallbacks to wrapped components 1`] = `
<Component>
  <SomeTestComponent
    kebab-label="Du, du hast."
    message="Eine Schweinshaxe, bitte."
    title="Good day, Sir."
  >
    <img
      width={100}
    />
    <h1>
      Good day, Sir.
    </h1>
    <h2>
      Du, du hast.
    </h2>
    <p>
      Eine Schweinshaxe, bitte.
    </p>
  </SomeTestComponent>
</Component>
`;
